pkg/registry/registry.go

```diff
@@ -0,0 +1,211 @@
+	manifestsDir, err := storeManifests(ctx, localManifestsDir)
+	if err != nil {
+		return nil, fmt.Errorf("storing manifests: %w", err)
+	}
+
+	charts, err := storeHelmCharts(ctx, helmClient)
+	if err != nil {
+		return nil, fmt.Errorf("storing helm charts: %w", err)
+	}
+
+	return &Registry{
+		embeddedImages: ctx.ImageDefinition.EmbeddedArtifactRegistry.ContainerImages,
+		manifestsDir:   manifestsDir,
+		helmClient:     helmClient,
+		helmCharts:     charts,
+		helmValuesDir:  helmValuesDir,
+		kubeVersion:    ctx.ImageDefinition.Kubernetes.Version,
+	}, nil
+}
+
+func (r *Registry) ManifestsPath() string {
+	return r.manifestsDir
+}
+
+func storeManifests(ctx *image.Context, localManifestsDir string) (string, error) {
+	var manifestsPathPopulated bool
+	manifestsDestDir := filepath.Join(ctx.BuildDir, "manifests")
+
+	manifestURLs := ctx.ImageDefinition.Kubernetes.Manifests.URLs
+	if len(manifestURLs) != 0 {
+		if err := os.MkdirAll(manifestsDestDir, os.ModePerm); err != nil {
+			return "", fmt.Errorf("creating manifests dir: %w", err)
+		}
+
+		for index, manifestURL := range manifestURLs {
+			filePath := filepath.Join(manifestsDestDir, fmt.Sprintf("dl-manifest-%d.yaml", index+1))
+
+			if err := http.DownloadFile(context.Background(), manifestURL, filePath, nil); err != nil {
+				return "", fmt.Errorf("downloading manifest '%s': %w", manifestURL, err)
+			}
+		}
+
+		manifestsPathPopulated = true
+	}
+
+	if _, err := os.Stat(localManifestsDir); err == nil {
+		if err = fileio.CopyFiles(localManifestsDir, manifestsDestDir, "", false); err != nil {
+			return "", fmt.Errorf("copying manifests: %w", err)
+		}
+
+		manifestsPathPopulated = true
+	} else if !errors.Is(err, fs.ErrNotExist) {
+		zap.S().Warnf("Searching for local manifests failed: %v", err)
+	}
+
+	if !manifestsPathPopulated {
+		return "", nil
+	}
+
+	return manifestsDestDir, nil
+}
+
+func storeHelmCharts(ctx *image.Context, helmClient helmClient) ([]*helmChart, error) {
+	helm := &ctx.ImageDefinition.Kubernetes.Helm
+
+	if len(helm.Charts) == 0 {
+		return nil, nil
+	}
+
+	helmDir := filepath.Join(ctx.BuildDir, "helm")
+	if err := os.MkdirAll(helmDir, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("creating helm directory: %w", err)
+	}
+
+	chartRepositories := mapChartRepos(helm)
+
+	var charts []*helmChart
+
+	for _, chart := range helm.Charts {
+		c := chart
+		repository, ok := chartRepositories[c.RepositoryName]
+		if !ok {
+			return nil, fmt.Errorf("repository not found for chart %s", c.Name)
+		}
+
+		localPath, err := downloadChart(helmClient, &c, repository, helmDir)
+		if err != nil {
+			return nil, fmt.Errorf("downloading chart: %w", err)
+		}
+
+		charts = append(charts, &helmChart{
+			HelmChart:     c,
+			localPath:     localPath,
+			repositoryURL: repository.URL,
+		})
+	}
+
+	return charts, nil
+}
+
+func mapChartRepos(helm *image.Helm) map[string]*image.HelmRepository {
```

Comment:
Nit: This is mostly covered by the return type, but `mapChartsToRepos` would be IMO a bit more clear.